barretenberg/cpp/src/barretenberg/avm_fuzzer/harness/calldata.fuzzer.cpp

```diff
@@ -0,0 +1,442 @@
+     *   - number of events (i.e. number of calldata instances to retrieve and hash)
+     *   - array of initial values (as in the memory gadget fuzzer, an array to fields to generate values from)
+     * Then for each event, we have a CalldataFuzzerInstance which configures:
+     *   - number of calldata fields
+     *   - selection encoding (as in the memory gadget fuzzer, configures generation of calldata values from the parent
+     * initial values)
+     *   - mutation (a choice of test case for this calldata instance):
+     *       - 0: do nothing to the calldata and emit as is
+     *       - 1: modify this calldata to be a copy of another instance
+     *       - 2: clear this calldata, so we emit an empty calldata array
+     *
+     * Every call to this custom mutator mutates **one** of:
+     * 0: starting context id
+     * 1: number of events
+     * 2: a single initial value
+     * 3: a single calldata instance
+     *
+     * If case 3 is chosen, one calldata instance is selected and **one** of the following is mutated for it:
+     * 0: mutation (choice of test case for this one calldata instance)
+     * 1: number of fields
+     * 3: selection encoding (how to generate the calldata fields)
+     *
+     * This method may be too 'nested' and granular, so it may be better to move to using something like
+     * mutate_calldata_vec rather than rely on initial values, where it is relatively slow to reach the case where we
+     * actually change the fields in a calldata instance.
+     */
+
+    switch (mutation_choice) {
+    case 0: {
+        // Modify number of events
+        std::uniform_int_distribution<uint8_t> num_events_dist(1, max_num_events);
+        input.num_events_input = num_events_dist(rng);
+        break;
+    }
+    case 1: {
+        // Modify initial context id
+        std::uniform_int_distribution<uint16_t> context_id_dist(
+            0, std::numeric_limits<uint16_t>::max() - input.num_events_input - 1);
+        input.start_context_id = context_id_dist(rng);
+        break;
+    }
+    case 2: {
+        // Modify a random initial value
+        // TODO(MW): Use mutate_calldata_vec (modify BASIC_VEC_MUTATION_CONFIGURATION for this fuzzer?)
+        std::uniform_int_distribution<size_t> index_dist(0, input.init_calldata_values.size() - 1);
+        size_t value_idx = index_dist(rng);
+        std::uniform_int_distribution<uint64_t> dist(0, std::numeric_limits<uint64_t>::max());
+        FF value = FF(dist(rng), dist(rng), dist(rng), dist(rng));
+        input.init_calldata_values[value_idx] = value;
+        break;
+    }
+    case 3: {
+        // Modify a random calldata instance (using num_events to ensure it's used in a run)
+        std::uniform_int_distribution<size_t> index_dist(0, input.num_events_input - 1);
+        size_t value_idx = index_dist(rng);
+        std::uniform_int_distribution<int> inner_mutation_dist(0, 2);
+        int inner_mutation_choice = inner_mutation_dist(rng);
```

Comment:
should probably break this out into a separate function, `mutate_calldata` so it's easier  to follow